LibraryImportGenerator does not respect C# 11 `scoped` modifier on parameters

### Description

A `scoped` modifier applied to a P/Invoke attributed with `[LibraryImport(...)]` is not replicated to the generated marshaling method.

### Reproduction Steps

```csharp
[LibraryImport("c", SetLastError = true)]
public static partial nint read(int fildes, scoped Span<byte> buf, nuint nbyte);
```


### Expected behavior

```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "7.0.6.47203")]
public static partial nint read(int fildes, scoped global::System.Span<byte> buf, nuint nbyte)
{
    // ...
}
```

### Actual behavior

```csharp
[System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Interop.LibraryImportGenerator", "7.0.6.47203")]
public static partial nint read(int fildes, global::System.Span<byte> buf, nuint nbyte)
{
    // ...
}
```

`error CS8988: The 'scoped' modifier of parameter 'buf' doesn't match partial method declaration.`

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

* .NET 7 RC 2
* Windows 11
* x64

### Other information

_No response_

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SignatureContext.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/SignatureContext.cs
@@ -12,7 +12,6 @@ using System.Threading;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Microsoft.CodeAnalysis.Diagnostics;
 using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
 
 namespace Microsoft.Interop
@@ -43,9 +42,24 @@ namespace Microsoft.Interop
                     if (typeInfo.ManagedIndex != TypePositionInfo.UnsetIndex
                         && typeInfo.ManagedIndex != TypePositionInfo.ReturnIndex)
                     {
+                        SyntaxTokenList tokens = TokenList();
+
+                        // "out" parameters are implicitly scoped, so we can't put the "scoped" keyword on them.
+                        // All other cases of explicit parameters are only scoped when the "scoped" keyword is present.
+                        // When the "scoped" keyword is present, it must be present on all declarations.
+                        if (typeInfo.ScopedKind != ScopedKind.None && typeInfo.RefKind != RefKind.Out)
+                        {
+                            tokens = tokens.Add(Token(SyntaxKind.ScopedKeyword));
+                        }
+
+                        if (typeInfo.IsByRef)
+                        {
+                            tokens = tokens.Add(Token(typeInfo.RefKindSyntax));
+                        }
+
                         yield return Parameter(Identifier(typeInfo.InstanceIdentifier))
                             .WithType(typeInfo.ManagedType.Syntax)
-                            .WithModifiers(TokenList(Token(typeInfo.RefKindSyntax)));
+                            .WithModifiers(tokens);
                     }
                 }
             }

--- a/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs
+++ b/src/libraries/System.Runtime.InteropServices/gen/Microsoft.Interop.SourceGeneration/TypePositionInfo.cs
@@ -66,6 +66,8 @@ namespace Microsoft.Interop
 
         public bool IsByRef => RefKind != RefKind.None;
 
+        public ScopedKind ScopedKind { get; init; } = ScopedKind.None;
+
         public ByValueContentsMarshalKind ByValueContentsMarshalKind { get; init; }
 
         public bool IsManagedReturnPosition { get => ManagedIndex == ReturnIndex; }
@@ -82,7 +84,8 @@ namespace Microsoft.Interop
                 InstanceIdentifier = ParseToken(paramSymbol.Name).IsReservedKeyword() ? $"@{paramSymbol.Name}" : paramSymbol.Name,
                 RefKind = paramSymbol.RefKind,
                 RefKindSyntax = RefKindToSyntax(paramSymbol.RefKind),
-                ByValueContentsMarshalKind = GetByValueContentsMarshalKind(paramSymbol.GetAttributes(), compilation)
+                ByValueContentsMarshalKind = GetByValueContentsMarshalKind(paramSymbol.GetAttributes(), compilation),
+                ScopedKind = paramSymbol.ScopedKind
             };
 
             return typeInfo;

--- a/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CodeSnippets.cs
+++ b/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/CodeSnippets.cs
@@ -503,6 +503,21 @@ namespace LibraryImportGenerator.UnitTests
         public static string ByValueParameterWithModifier<T>(string attributeName, string preDeclaration = "") => ByValueParameterWithModifier(typeof(T).ToString(), attributeName, preDeclaration);
 
         /// <summary>
+        /// Declaration with one parameter with custom modifiers.
+        /// </summary>
+        public static string SingleParameterWithModifier(string typeName, string modifiers, string preDeclaration = "") => $$"""
+            using System.Runtime.InteropServices;
+            using System.Runtime.InteropServices.Marshalling;
+            {{preDeclaration}}
+            partial class Test
+            {
+                [LibraryImport("DoesNotExist")]
+                public static partial void Method(
+                    {{modifiers}} {{typeName}} {|#0:p|});
+            }
+            """;
+
+        /// <summary>
         /// Declaration with by-value parameter with custom name.
         /// </summary>
         public static string ByValueParameterWithName(string methodName, string paramName) => $$"""

--- a/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/Compiles.cs
+++ b/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/Compiles.cs
@@ -248,6 +248,9 @@ namespace LibraryImportGenerator.UnitTests
 
             // Type-level interop generator trigger attributes
             yield return new[] { ID(), CodeSnippets.GeneratedComInterface };
+
+            // Parameter modifiers
+            yield return new[] { ID(), CodeSnippets.SingleParameterWithModifier("int", "scoped ref") };
         }
 
         public static IEnumerable<object[]> CustomCollections()

